sno/sno_repo.py

```diff
@@ -0,0 +1,146 @@
+# Causes all git commands that would involve the index or working copy to fail with "unsupported extension: .sno" -
+# in that sense it is "locked" to git. Various techniques can be used to unlock it if certain git functionality is
+# needed - eg marking the repository as bare so it is ignored, or removing the unsupported extension.
+
+# See https://git-scm.com/docs/index-format
+LOCKED_EMPTY_GIT_INDEX = _append_checksum(EMPTY_GIT_INDEX + LOCKED_SNO_EXTENSION)
+
+
+def init_repository(repo_root_path, repo_version):
+    """
+    Initialise a new sno repo. A sno repo is basically a git repo, except -
+    - git internals are stored in .sno instead of .git
+      (.git is a file that contains a reference to .sno, this is allowed by git)
+    - datasets are stored in /.sno-dataset/ trees according to a particular dataset format version - see DATASETS_v2.md
+      but, this only matters once we start committing data. When initialising a repo, these are not yet present.
+    - there is a blob called sno.repository.version that contains the dataset format version number - but, this is only
+      written when data is first imported. When initialising a repo, this is not yet present.
+    - there is property in the repo config called sno.repository.version that contains the dataset format version
+      number, which is used until the sno.repository.version blob is written.
+    - the .sno/index file has been extended to stop git messing things up - see LOCKED_EMPTY_GIT_INDEX.
+
+    There can also be properties in the repo config relating to the working copy - named sno.workingcopy.* - but
+    these are not written by this function, it is up to the caller to configure these as needed.
+    See WorkingCopy.write_config()
+    """
+
+    repo = pygit2.init_repository(str(repo_root_path.resolve()))
+    repo = convert_git_repo_to_sno_repo(repo_root_path)
+    write_repo_version_config(repo, repo_version)
+    assert is_sno_repo(repo)
+    return repo
+
+
+def convert_git_repo_to_sno_repo(repo_root_path):
+    """Given a newly created non-bare git repo, with the following structure:
+    repo_root_path/  # also worktree path
+        .git/
+            HEAD
+            ...      # objects, refs, etc
+    Converts it to a sno repo with the following structure:
+    repo_root_path/  # also worktree path, as before
+        .git         # contains "gitdir: .sno"
+        .sno/
+            HEAD
+            index    # Locked using .sno extension
+            ...      # etc
+    """
+
+    # Note: pygit2 can read the .sno directory okay, but it won't create it, so we do it ourselves using Path.rename()
+    # Note: We could also call `git init --separate-git-dir=.sno` to crete it but it doesn't work very well - it
+    # converts all paths to be absolute, which would mean sno repos wouldn't be renameable.
+
+    # Move .git to .sno but add a reference at .git so git can find it.
+    dot_git_path = repo_root_path / ".git"
+    assert dot_git_path.is_dir()
```

Comment:
`assert` statements are a bit of a danger in production Python code, since `python -O` compiles them away, is better to conditionally raise an `AssertionError` explicitly.